mypy/checkexpr.py

```diff
@@ -1304,7 +1304,9 @@ def infer_overload_return_type(self,
             return None
         elif any_causes_overload_ambiguity(matches, return_types, arg_types, arg_kinds, arg_names):
             # An argument of type or containing the type 'Any' caused ambiguity.
-            if all(is_subtype(ret_type, return_types[-1]) for ret_type in return_types[:-1]):
+            if all(is_subtype(ret_type, return_types[-1]) and
+                   is_subtype(return_types[-1], ret_type)
+                   for ret_type in return_types[:-1]):
                 # The last match is a supertype of all the previous ones, so it's safe
                 # to return that inferred type.
```

Comment:
We should probably update this comment. Other then that, lgtm

Reply:
OK, will do.